File: src/com/anathema_roguelike/characters/player/perks/specializations/AbilitySpecializationChoice.java
package com.anathema_roguelike.characters.player.perks.specializations;

import java.util.Collection;
import java.util.stream.Collectors;

import com.anathema_roguelike.characters.Character;
import com.anathema_roguelike.characters.perks.Perk;
import com.anathema_roguelike.characters.perks.PerkChoice;
import com.anathema_roguelike.characters.player.perks.abilities.Ability;
import com.anathema_roguelike.main.utilities.Utils;
import com.google.common.collect.Collections2;

public class AbilitySpecializationChoice extends PerkChoice {
	
	private Class<? extends Ability> ability;
	
	public AbilitySpecializationChoice(Class<? extends Ability> ability) {
		super("Choose an Ability to specialize into");
		
		this.ability = ability;
	}

	@Override
	public Collection<? extends Perk> getChoices(Character character) {
		
		Collection<? extends Ability> learnedAbilities = Utils.filterBySubclass(character.getPerks(Perk.class), Ability.class);
		
		return Collections2.filter(learnedAbilities, a -> validAbility(character, a))
				.stream().map(a -> new AbilitySpecialization(a.getClass()))
				.collect(Collectors.toList());
	}
	
	public boolean validAbility(Character character, Ability a) {
		return ability.isAssignableFrom(a.getClass()) && getCharacter().getSpecialization(a.getClass()) < 3;
	}
}

File: src/com/anathema_roguelike/entities/characters/player/perks/specializations/AbilitySpecializationChoice.scala
/*******************************************************************************
 * Copyright (c) 2019. Carl Minden
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/

package com.anathema_roguelike
package entities.characters.player.perks.specializations

import com.anathema_roguelike.entities.characters.perks.Perk
import com.anathema_roguelike.entities.characters.perks.PerkChoice
import com.anathema_roguelike.entities.characters.player.perks.abilities.Ability
import scala.reflect.runtime.universe._

import com.anathema_roguelike.entities.characters.Character

class AbilitySpecializationChoice[T <: Ability : TypeTag] extends PerkChoice("Choose an Ability to specialize into") {
  def getChoices(character: Character): Iterable[_ <: Perk] = {

    val learnedAbilities = character.getPerks[Perk].filterByType[T]

    learnedAbilities.filter(a => validAbility(character, a)).map(a => new AbilitySpecialization()(a.getClass))
  }

  def validAbility(character: Character, a: Ability): Boolean = {
    typeTagToClass[T].isAssignableFrom(a.getClass) && character.getSpecialization(a.getClass) < 3
  }
}